src/mygrad/operation_base.py

```diff
@@ -152,9 +152,13 @@ def backward(self, grad, *, graph, _reduction=None, **kwargs):
                         var.grad += backed_grad
                     else:
                         var.grad += _reduction(backed_grad, var.shape)
+        # Avoid visiting the same node multiple times. Note that we don't store
+        # these by the node itself, since Tensors are unhashable, but by its `id`.
+        visited = set()
         for var in (
-            i for i in self.variables if not i.constant and i.creator is not None
+            i for i in self.variables if not i.constant and i.creator is not None and id(i) not in visited
```

Comment:
I think this breaks `x * x` <- the same tensor is used twice. If the test suite doesn't catch this, then I definitely need to add tests for it.

Reply:
They don't! Everything's passing locally. Why would this break it? It seems equivalent to the world before.